openedx/features/content_type_gating/partitions.py

```diff
@@ -0,0 +1,192 @@
+        # partition with id 51, it will collide with the Content Gating Partition. We'll catch that here, and
+        # then fix the course content as needed (or get the course team to).
+        LOG.warning(
+            "Can't add 'content_type_gate' partition, as ID {id} is assigned to {partition} in course {course}.".format(
+                id=CONTENT_GATING_PARTITION_ID,
+                partition=_get_partition_from_id(course.user_partitions, CONTENT_GATING_PARTITION_ID).name,
+                course=unicode(course.id)
+            )
+        )
+        return None
+
+    partition = content_gate_scheme.create_user_partition(
+        id=CONTENT_GATING_PARTITION_ID,
+        name=_(u"Feature-based Enrollments"),
+        description=_(u"Partition for segmenting users by access to gated content types"),
+        parameters={"course_id": unicode(course.id)}
+    )
+    return partition
+
+
+class ContentTypeGatingPartition(UserPartition):
+    pass
+
+
+class ContentTypeGatingPartitionScheme(object):
+    """
+    This scheme implements the Content Type Gating permission partitioning.
+
+    This partitioning is roughly the same as the verified/audit split, but also allows for individual
+    schools or courses to specify particular learner subsets by email that are allowed to access
+    the gated content despite not being verified users.
+    """
+
+    LIMITED_ACCESS = Group(settings.CONTENT_TYPE_GATE_GROUP_IDS['limited_access'], 'Limited-access Users')
+    FULL_ACCESS = Group(settings.CONTENT_TYPE_GATE_GROUP_IDS['full_access'], 'Full-access Users')
+
+    @classmethod
+    def get_group_for_user(cls, course_key, user, user_partition, **kwargs):
+        """
+        Returns the Group for the specified user.
+        """
+
+        # First, check if we have to deal with masquerading.
+        # If the current user is masquerading as a specific student, use the
+        # same logic as normal to return that student's group. If the current
+        # user is masquerading as a generic student in a specific group, then
+        # return that group.
+        if get_course_masquerade(user, course_key) and not is_masquerading_as_specific_student(user, course_key):
+            return get_masquerading_user_group(course_key, user, user_partition)
+
+        # For now, treat everyone as a Full-access user, until we have the rest of the
+        # feature gating logic in place.
+
+        if not CONTENT_TYPE_GATING_FLAG.is_enabled():
+            return cls.FULL_ACCESS
+
+        # If CONTENT_TYPE_GATING is enabled use the following logic to determine whether a user should have FULL_ACCESS
+        # or LIMITED_ACCESS
+
+        course_mode = apps.get_model('course_modes.CourseMode')
+        modes = course_mode.modes_for_course_dict(course_key)
+
+        # If there is no verified mode, all users are granted FULL_ACCESS
+        if not course_mode.has_verified_mode(modes):
```

Comment:
is this reflected in our product document?  
it would be valuable to add it if it is not

Reply:
Good point - I'll make sure that is added